# EDA -> Exploratory Data Analysis
Exploring the <a href="https://www.kaggle.com/datasets/parulpandey/2020-it-salary-survey-for-eu-region">**IT Salary Survey for EU region**</a> dataset.
### Dataset Info
- **Contex**
      An anonymous salary survey has been conducted annually since 2015 among European IT specialists with a stronger focus on Germany. This year 1238 respondents volunteered to participate in the survey. The data has been made publicly available by the authors.
      the dataset contains rich information about the salary patterns among the IT professionals in the EU region and offers some great insghts.
- **File**
      - <a href="https://www.kaggle.com/datasets/parulpandey/2020-it-salary-survey-for-eu-region?select=IT+Salary+Survey+EU++2018.csv">IT Salary Survey EU 2018.csv</a>
            &ensp;&ensp;Row&ensp;&ensp;&ensp;&ensp;: 765 rows
            &ensp;&ensp;Column&ensp;: 14 columns
      - <a href="https://www.kaggle.com/datasets/parulpandey/2020-it-salary-survey-for-eu-region?select=IT+Salary+Survey+EU++2019.csv">IT Salary Survey EU 2019.csv</a>
            &ensp;&ensp;Row&ensp;&ensp;&ensp;&ensp;: 991 rows
            &ensp;&ensp;Column&ensp;: 23 columns
      - <a href="https://www.kaggle.com/datasets/parulpandey/2020-it-salary-survey-for-eu-region?select=IT+Salary+Survey+EU++2020.csv">IT Salary Survey EU 2020.csv</a>
            &ensp;&ensp;Row&ensp;&ensp;&ensp;&ensp;: 1253 rows
            &ensp;&ensp;Column&ensp;: 23 columns

In [1]:
# Importing the used dataset
import pandas as pd
import os

In [2]:
# Declaring the path as global variable
PATH_MAIN = r"./../Dataset/"
PATH_18 = os.path.join(PATH_MAIN, r"IT Salary Survey EU 2018.csv")
PATH_19 = os.path.join(PATH_MAIN, r"IT Salary Survey EU 2019.csv")
PATH_20 = os.path.join(PATH_MAIN, r"IT Salary Survey EU 2020.csv")

In [3]:
# Getting the dataset (csv file) into the script
data_18 = pd.read_csv(PATH_18)
data_19 = pd.read_csv(PATH_19)
data_20 = pd.read_csv(PATH_20)

## Step 1 - Observe The Overview
Getting the general overview of dataset

In [6]:
# Getting the columns name from each file
column_18 = data_18.columns
column_19 = data_19.columns
column_20 = data_20.columns
print(f'== 2018 Data == [{len(column_18)} columns]\n{column_18}',
      f'== 2019 Data == [{len(column_19)} columns]\n{column_19}',
      f'== 2020 Data == [{len(column_20)} columns]\n{column_20}',
      sep='\n\n')

== 2018 Data == [14 columns]
Index(['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years of experience',
       'Your level', 'Current Salary', 'Salary one year ago',
       'Salary two years ago', 'Are you getting any Stock Options?',
       'Main language at work', 'Company size', 'Company type'],
      dtype='object')

== 2019 Data == [23 columns]
Index(['Zeitstempel', 'Age', 'Gender', 'City', 'Seniority level',
       'Position (without seniority)', 'Years of experience',
       'Your main technology / programming language',
       'Yearly brutto salary (without bonus and stocks)', 'Yearly bonus',
       'Yearly stocks',
       'Yearly brutto salary (without bonus and stocks) one year ago. Only answer if staying in same country',
       'Yearly bonus one year ago. Only answer if staying in same country',
       'Yearly stocks one year ago. Only answer if staying in same country',
       'Number of vacation days', 'Number of home office days per month',
       'Main language at 

#### Insight
- **Data 2018**
      This table had 14 columns
- **Data 2019**
      This table had 23 columns
- **Data 2020**
      This table had 23 columns

### 2018 Dataset

In [7]:
# 2018 data sneakpeak
data_18.head(5)

Timestamp   Age Gender     City                       Position  \
0  14/12/2018 12:41:33  43.0      M  München                   QA Ingenieur   
1  14/12/2018 12:42:09  33.0      F  München  Senior PHP Magento developer    
2  14/12/2018 12:47:36  32.0      M  München              Software Engineer   
3  14/12/2018 12:50:15  25.0      M  München      Senior Frontend Developer   
4  14/12/2018 12:50:31  39.0      M  München                    UX Designer   

   Years of experience Your level  Current Salary  Salary one year ago  \
0                 11.0     Senior         77000.0              76200.0   
1                  8.0     Senior         65000.0              55000.0   
2                 10.0     Senior         88000.0              73000.0   
3                  6.0     Senior         78000.0              55000.0   
4                 10.0     Senior         69000.0              60000.0   

   Salary two years ago Are you getting any Stock Options?  \
0               68000.0                                 No   
1               55000.0                                 No   
2               54000.0                                 No   
3               45000.0                                Yes   
4               52000.0                                 No   

  Main language at work Company size   Company type  
0               Deutsch     100-1000        Product  
1               Deutsch       50-100        Product  
2               Deutsch        1000+        Product  
3               English        1000+        Product  
4               English     100-1000  Ecom retailer

In [17]:
data_18.shape

(765, 14)

**Insight**
The data has 765 row with 14 columns

In [10]:
data_18.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 765 entries, 0 to 764
Data columns (total 14 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Timestamp                           765 non-null    object 
 1   Age                                 672 non-null    float64
 2   Gender                              751 non-null    object 
 3   City                                736 non-null    object 
 4   Position                            737 non-null    object 
 5   Years of experience                 732 non-null    float64
 6   Your level                          743 non-null    object 
 7   Current Salary                      750 non-null    float64
 8   Salary one year ago                 596 non-null    float64
 9   Salary two years ago                463 non-null    float64
 10  Are you getting any Stock Options?  742 non-null    object 
 11  Main language at work               750 non-n

In [28]:
temp_18 = data_18.dtypes.to_frame().reset_index()
temp_18.rename(columns={"index": "Column", 0: "Dtype"}, inplace=True)
temp_18[temp_18.Dtype == 'float64']

Column    Dtype
1                   Age  float64
5   Years of experience  float64
7        Current Salary  float64
8   Salary one year ago  float64
9  Salary two years ago  float64

In [29]:
temp_18[temp_18.Dtype == 'object']

Column   Dtype
0                            Timestamp  object
2                               Gender  object
3                                 City  object
4                             Position  object
6                           Your level  object
10  Are you getting any Stock Options?  object
11               Main language at work  object
12                        Company size  object
13                        Company type  object

**Insight**
The data has 14 columns with almost all the column except Timestamp column has null value. The column data type consist of 5 float64 column and 9 object column.

In [23]:
data_18["Gender"].value_counts()

Gender
M    646
F    105
Name: count, dtype: int64

**Insight**
Column Gender has 2 values, *M* for *Male* and *F* for *Female*. There are 646 male and 105 female worker respondent.

In [24]:
data_18["City"].value_counts()

City
Berlin         291
München        249
Frankfurt       33
Köln            20
Hamburg         16
              ... 
Gdansk           1
Poland           1
Düsseldorf       1
Ludwigsburg      1
Andernach        1
Name: count, Length: 74, dtype: int64

**Insight**
The respondent come from 74 different cities with Berlin and Munchen being the most.

In [25]:
data_18["Age"].min()

21.0

In [26]:
data_18["Age"].max()

60.0

**Insight**
The respondent age is range from 21 to 60 years old

In [8]:
# 2019 data sneakpeak
data_19.head(5)

Zeitstempel   Age Gender    City Seniority level  \
0  02.12.2019 11:18:26  33.0   Male  Berlin          Senior   
1  02.12.2019 11:18:35  29.0   Male  Berlin          Middle   
2  02.12.2019 11:18:56   NaN   Male  Berlin          Middle   
3  02.12.2019 11:19:08  30.0   Male  Berlin          Senior   
4  02.12.2019 11:19:37  32.0   Male  Berlin          Senior   

  Position (without seniority)  Years of experience  \
0          Fullstack Developer                   13   
1            Backend Developer                    3   
2             Mobile Developer                    4   
3            Backend Developer                    6   
4           Embedded Developer                   10   

  Your main technology / programming language  \
0                                         PHP   
1                                      Python   
2                                      Kotlin   
3                                         PHP   
4                                       C/C++   

   Yearly brutto salary (without bonus and stocks)  Yearly bonus  ...  \
0                                          64000.0        1000.0  ...   
1                                          55000.0           NaN  ...   
2                                          70000.0           NaN  ...   
3                                          63000.0           NaN  ...   
4                                          66000.0           NaN  ...   

   Yearly stocks one year ago. Only answer if staying in same country  \
0                                                NaN                    
1                                                NaN                    
2                                                NaN                    
3                                                NaN                    
4                                                NaN                    

   Number of vacation days  Number of home office days per month  \
0                     29.0                                   4.0   
1                     22.0                                   4.0   
2                     27.0                                   NaN   
3                     24.0                                   NaN   
4                     30.0                                   0.0   

   Main language at work  Company name   Company size Company type  \
0                English            NaN        50-100      Startup   
1                English            NaN         10-50      Product   
2                English            NaN         1000+      Startup   
3                English          Auto1      100-1000      Product   
4                English         Luxoft        50-100      Product   

  Сontract duration Company business sector   0  
0         unlimited                 Tourism NaN  
1         unlimited   Scientific Activities NaN  
2         unlimited                     NaN NaN  
3         unlimited               Transport NaN  
4         unlimited              Automotive NaN  

[5 rows x 23 columns]

In [27]:
data_19.shape

(991, 23)

**Insight**
The data has 23 column with 991 row

In [11]:
data_19.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 991 entries, 0 to 990
Data columns (total 23 columns):
 #   Column                                                                                                Non-Null Count  Dtype  
---  ------                                                                                                --------------  -----  
 0   Zeitstempel                                                                                           991 non-null    object 
 1   Age                                                                                                   882 non-null    float64
 2   Gender                                                                                                991 non-null    object 
 3   City                                                                                                  991 non-null    object 
 4   Seniority level                                                                                       97

In [30]:
temp_19 = data_19.dtypes.to_frame().reset_index()
temp_19.rename(columns={"index": "Column", 0: "Dtype"}, inplace=True)
temp_19[temp_19.Dtype == 'float64']

Column    Dtype
1                                                 Age  float64
8     Yearly brutto salary (without bonus and stocks)  float64
9                                        Yearly bonus  float64
10                                      Yearly stocks  float64
11  Yearly brutto salary (without bonus and stocks...  float64
12  Yearly bonus one year ago. Only answer if stay...  float64
13  Yearly stocks one year ago. Only answer if sta...  float64
14                            Number of vacation days  float64
15               Number of home office days per month  float64
22                                                  0  float64

In [32]:
temp_19[temp_19.Dtype == "int64"]

Column  Dtype
6  Years of experience  int64

In [31]:
temp_19[temp_19.Dtype == 'object']

Column   Dtype
0                                   Zeitstempel  object
2                                        Gender  object
3                                          City  object
4                               Seniority level  object
5                  Position (without seniority)  object
7   Your main technology / programming language  object
16                        Main language at work  object
17                                Company name   object
18                                 Company size  object
19                                 Company type  object
20                            Сontract duration  object
21                      Company business sector  object

**Insight**
The data has 23 column with only 4 column doesn't have null value. The column consist of 1 int64 type, 10 float64 type, and 12 object type column.

In [33]:
data_19["Gender"].value_counts()

Gender
Male      838
Female    153
Name: count, dtype: int64

**Insight**
There are 838 male and 153 female respondent

In [34]:
data_19["City"].value_counts()

City
Berlin       430
Munich       240
Amsterdam     80
Frankfurt     50
Hamburg       34
            ... 
Jyvaskyla      1
Toulouse       1
Dubai          1
Lingen         1
Kiev           1
Name: count, Length: 76, dtype: int64

**Insight**
The respondent came from 76 different cities with Berlin and Munich as the most.

In [36]:
data_19["Age"].min()

20.0

In [37]:
data_19["Age"].max()

54.0

**Insight**
The respondent age is ranged from 20 to 54 years old.

In [9]:
# 2020 data sneakpeak
data_20.head(5)

Timestamp   Age Gender    City           Position   \
0  24/11/2020 11:14:15  26.0   Male  Munich   Software Engineer   
1  24/11/2020 11:14:16  26.0   Male  Berlin   Backend Developer   
2  24/11/2020 11:14:21  29.0   Male  Berlin   Software Engineer   
3  24/11/2020 11:15:24  28.0   Male  Berlin  Frontend Developer   
4  24/11/2020 11:15:46  37.0   Male  Berlin   Backend Developer   

  Total years of experience Years of experience in Germany Seniority level  \
0                         5                              3          Senior   
1                         7                              4          Senior   
2                        12                              6            Lead   
3                         4                              1          Junior   
4                        17                              6          Senior   

  Your main technology / programming language  \
0                                  TypeScript   
1                                        Ruby   
2                     Javascript / Typescript   
3                                  Javascript   
4                                     C# .NET   

  Other technologies/programming languages you use often  ...  \
0                    Kotlin, Javascript / Typescript      ...   
1                                                NaN      ...   
2                    Javascript / Typescript, Docker      ...   
3                                                NaN      ...   
4                             .NET, SQL, AWS, Docker      ...   

   Annual bonus+stocks one year ago. Only answer if staying in same country  \
0                                              10000                          
1                                               5000                          
2                                             100000                          
3                                                NaN                          
4                                                NaN                          

  Number of vacation days           Employment status   Сontract duration  \
0                      30          Full-time employee  Unlimited contract   
1                      28          Full-time employee  Unlimited contract   
2                      30  Self-employed (freelancer)  Temporary contract   
3                      24          Full-time employee  Unlimited contract   
4                      29          Full-time employee  Unlimited contract   

  Main language at work Company size Company type  \
0               English       51-100      Product   
1               English     101-1000      Product   
2               English     101-1000      Product   
3               English       51-100      Startup   
4               English     101-1000      Product   

  Have you lost your job due to the coronavirus outbreak?  \
0                                                 No        
1                                                 No        
2                                                Yes        
3                                                 No        
4                                                 No        

  Have you been forced to have a shorter working week (Kurzarbeit)? If yes, how many hours per week  \
0                                                NaN                                                  
1                                                NaN                                                  
2                                                NaN                                                  
3                                                NaN                                                  
4                                                NaN                                                  

  Have you received additional monetary support from your employer due to Work From Home? If yes, how much in 2020 in EUR  
0                                                NaN                                    

In [38]:
data_20.shape

(1253, 23)

**Insight**
The data has 23 column with 1253 row/respondent

In [12]:
data_20.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1253 entries, 0 to 1252
Data columns (total 23 columns):
 #   Column                                                                                                                   Non-Null Count  Dtype  
---  ------                                                                                                                   --------------  -----  
 0   Timestamp                                                                                                                1253 non-null   object 
 1   Age                                                                                                                      1226 non-null   float64
 2   Gender                                                                                                                   1243 non-null   object 
 3   City                                                                                                                     1253 non-null   o

In [39]:
temp_20 = data_20.dtypes.to_frame().reset_index()
temp_20.rename(columns={"index": "Column", 0: "Dtype"}, inplace=True)
temp_20[temp_20.Dtype == 'float64']

Column    Dtype
1                                                 Age  float64
10  Yearly brutto salary (without bonus and stocks...  float64
12  Annual brutto salary (without bonus and stocks...  float64
21  Have you been forced to have a shorter working...  float64

In [40]:
temp_20[temp_20.Dtype == 'object']

Column   Dtype
0                                           Timestamp  object
2                                              Gender  object
3                                                City  object
4                                           Position   object
5                           Total years of experience  object
6                      Years of experience in Germany  object
7                                     Seniority level  object
8         Your main technology / programming language  object
9   Other technologies/programming languages you u...  object
11                       Yearly bonus + stocks in EUR  object
13  Annual bonus+stocks one year ago. Only answer ...  object
14                            Number of vacation days  object
15                                  Employment status  object
16                                  Сontract duration  object
17                              Main language at work  object
18                                       Company size  object
19                                       Company type  object
20  Have you lost your job due to the coronavirus ...  object
22  Have you received additional monetary support ...  object

**Insight**
The data has 23 column with only 3 column doesn't have null value. The 23 column consist of 19 object type and 4 float64 type column.

In [13]:
temp_dict_column = {}
for item18 in column_18:
      temp_dict_column[item18] = 1

for item19 in column_19:
      if item19 in temp_dict_column.keys():
            temp_dict_column[item19] += 1
      else:
            temp_dict_column[item19] = 1

for item20 in column_20:
      if item20 in temp_dict_column.keys():
            temp_dict_column[item20] += 1
      else:
            temp_dict_column[item20] = 1

matching_column = []
for key, value in temp_dict_column.items():
      if value == 3:
            matching_column.append(key)

print("matching column in data 2018-2020 is:", matching_column)

matching column in data 2018-2020 is: ['Age', 'Gender', 'City', 'Main language at work', 'Company size', 'Company type']
